src/System.Linq.Expressions/tests/Interpreter/InterpreterTests.Call.cs

```diff
@@ -0,0 +1,137 @@
+
+#if FEATURE_INTERPRET
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using Xunit;
+
+namespace Tests.Expressions
+{
+    partial class InterpreterTests
+    {
+        [Fact(Skip = "4150")]
+        public static void CompileInterpretCrossCheck_Call_WriteBacks()
+        {
+            foreach (var e in Call_WriteBacks())
+            {
+                Verify(e);
+            }
+        }
+
+        private static IEnumerable<Expression> Call_WriteBacks()
+        {
+            // OK
+            foreach (var t in new[] { typeof(C), typeof(S) })
+            {
+                var p = Expression.Parameter(t);
+                var m = Expression.PropertyOrField(p, "X");
+                var mtd = t.GetTypeInfo().GetDeclaredMethod("Do");
+
+                yield return Expression.Block(new[] { p }, Expression.Assign(p, Expression.New(t)), Expression.Call(p, mtd), m);
+            }
+
+            // OK
+            foreach (var t in new[] { typeof(C), typeof(S) })
+            {
+                var p = Expression.Parameter(t.MakeArrayType());
+                var o = Expression.ArrayIndex(p, Expression.Constant(0));
+                var m = Expression.PropertyOrField(o, "X");
+                var mtd = t.GetTypeInfo().GetDeclaredMethod("Do");
+
+                yield return Expression.Block(new[] { p }, Expression.Assign(p, Expression.NewArrayInit(t, Expression.New(t))), Expression.Call(o, mtd), m);
+            }
+
+            // OK
+            foreach (var t in new[] { typeof(C), typeof(S) })
+            {
+                var p = Expression.Parameter(t.MakeArrayType(2));
+                var o = Expression.ArrayAccess(p, Expression.Constant(0), Expression.Constant(0));
+                var m = Expression.PropertyOrField(o, "X");
+                var mtd = t.GetTypeInfo().GetDeclaredMethod("Do");
+
+                yield return Expression.Block(new[] { p }, Expression.Assign(p, Expression.NewArrayBounds(t, Expression.Constant(1), Expression.Constant(1))), Expression.Assign(o, Expression.New(t)), Expression.Call(o, mtd), m);
+            }
+
+            // FAIL - Issue for struct; compiler doesn't write back but interpreter does
+            foreach (var t in new[] { typeof(C), typeof(S) })
+            {
+                var p = Expression.Parameter(typeof(Holder<>).MakeGenericType(t));
+                var o = Expression.Property(p, "Value");
+                var m = Expression.PropertyOrField(o, "X");
+                var mtd = t.GetTypeInfo().GetDeclaredMethod("Do");
+
+                yield return Expression.Block(new[] { p }, Expression.Assign(p, Expression.New(p.Type.GetTypeInfo().DeclaredConstructors.Single(), Expression.New(t))), Expression.Call(o, mtd), m);
+            }
+
+            // OK - Surprising compared to previous
+            foreach (var t in new[] { typeof(C), typeof(S) })
```

Comment:
Not really surprising. The compiler will use `ldflda` and `call` when the type of an instance to a call is a value-type, which among other things means that mutable structs get mutated in place (it's also required for the virtual methods of value-types to be verifiable). It's presumably this behaviour that the write-back of the interpreter is meant to emulate.
In the previous case, it would have used `ldloca` against a local that was previously obtained with the call to the property, so only that local is mutated in place, and then the mutation lost. (This confusing people is one of the reasons for mutable structs being declared evil by many, when they're not evil, they're just misunderstood, which is worse).

You can see the same thing in the C# of:
```C#
var holder = new Holder<List<int>.Enumerator>(new List<int>{1}.GetEnumerator());
Assert.True(holder.Value.MoveNext());
Assert.True(holder.Value.MoveNext());
holder = new Holder<List<int>.Enumerator>(new List<int>{1}.GetEnumerator());
holder.Value = new List<int>{1}.GetEnumerator();
Assert.True(holder._value.MoveNext());
Assert.False(holder._value.MoveNext());